packages/website/src/components/ast/serializer/serializerScope.ts

```diff
@@ -0,0 +1,203 @@
+import type { ASTViewerModel, Serializer, SelectedRange } from '../types';
+import type { TSESTree } from '@typescript-eslint/website-eslint';
+import { isRecord } from '../utils';
+
+function isESTreeNode(
+  value: unknown,
+): value is Record<string, unknown> & TSESTree.Node {
+  return Boolean(value) && isRecord(value) && 'type' in value && 'loc' in value;
+}
+
+function getClassName(value: Record<string, unknown>): string {
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  return (Object.getPrototypeOf(value) as Object).constructor.name.replace(
+    /\$[0-9]+$/,
+    '',
+  );
+}
+
+function getNodeName(
+  className: string,
+  data: Record<string, unknown>,
+): string | undefined {
+  const id = data.$id != null ? `$${String(data.$id)}` : '';
+
+  if (className === 'ImplicitLibVariable' && data.name === 'const') {
+    className = 'ImplicitGlobalConstTypeVariable';
+  }
+
+  return `${className}${id}`;
+}
+
+function getRange(value: Record<string, unknown>): SelectedRange | undefined {
+  if (isESTreeNode(value.block)) {
+    return {
+      start: value.block.loc.start,
+      end: value.block.loc.end,
+    };
+  } else if (isESTreeNode(value.identifier)) {
+    return {
+      start: { ...value.identifier.loc.start },
+      end: { ...value.identifier.loc.end },
+    };
+  } else if (isESTreeNode(value.node)) {
+    return {
+      start: { ...value.node.loc.start },
+      end: { ...value.node.loc.end },
+    };
+  } else if (
+    Array.isArray(value.identifiers) &&
+    value.identifiers.length > 0 &&
+    isESTreeNode(value.identifiers[0])
+  ) {
+    return {
+      start: { ...value.identifiers[0].loc.start },
+      end: { ...value.identifiers[0].loc.end },
+    };
+  }
+
+  return undefined;
+}
+
+type NodeType =
+  | 'Scope'
+  | 'Definition'
+  | 'Variable'
+  | 'ScopeManager'
+  | 'Reference';
+
+function getNodeType(nodeName: string | undefined): NodeType | undefined {
+  if (nodeName) {
+    if (nodeName === 'ScopeManager') {
+      return 'ScopeManager';
+    } else if (nodeName.endsWith('Scope')) {
+      return 'Scope';
+    } else if (nodeName.endsWith('Definition')) {
+      return 'Definition';
+    } else if (nodeName === 'Variable' || nodeName === 'ImplicitLibVariable') {
+      return 'Variable';
+    } else if (nodeName === 'Reference') {
+      return 'Reference';
+    }
+  }
+  return undefined;
+}
+
+function getProps(nodeType: NodeType | undefined): string[] | undefined {
+  switch (nodeType) {
+    case 'ScopeManager':
+      return ['scopes', 'globalScope', 'variables'];
+    case 'Scope':
+      return [
+        'block',
+        'isStrict',
+        'references',
+        'through',
+        'set',
+        'type',
+        'variables',
+        'variableScope',
+        'functionExpressionScope',
+        'childScopes',
+        'upper',
+      ];
```

Comment:
nitpick - to ensure we don't accidentally add bad keys here or forget to remove keys - would this work to validate the keys are correct?

(if it does - we should do it for all of the arrays here)

```suggestion
      ] as (keyof Scope)[];
```

Reply:
there is runtime validation if field is present, not sure if we want to cast this to `Scope`

Reply:
if we want to ensure that all properties are printed, we could remove this function completely and instead of

```ts
      const props = getProps(nodeType);
      if (props) {
        values = props.map(key => {
          const res = data[key];
          return [key, typeof res === 'function' ? res.bind(data) : res];
        });
      } else {
        values = Object.entries(data);
      }
```

do
```ts
      values = Object.entries(nodeType).map(([key, res]) => {x
         return [key, typeof res === 'function' ? res.bind(data) : res];
      });
```

Reply:
as for functions i'm going to add in new PR support for lazy execution for typescript types